File: 2444-count-subarrays-with-fixed-bounds/2444-count-subarrays-with-fixed-bounds.py
class Solution:
    def countSubarrays(self, nums: List[int], minK: int, maxK: int) -> int:
        ans = 0
		# kmin: number of subarrays ending at last pos that contains minK
		# kmax: number of subarrays ending at last pos that contains maxK
		# start: pos of previous invalid number (namely > maxK or < minK)
        kmin, kmax, start = 0, 0, -1
        for i, num in enumerate(nums):
            if num < minK or num > maxK:
                kmin, kmax, start = 0, 0, i
            else:
                if num == minK:
                    kmin = i - start
                if num == maxK:
                    kmax = i - start            
                ans += min(kmin, kmax)
        return ans
            
                
                
        
                    
                

        
        
        
        
        
        
        

File: 2444-count-subarrays-with-fixed-bounds/2444-count-subarrays-with-fixed-bounds.java
class Solution {
    public long countSubarrays(int[] nums, int minK, int maxK) {

        long res = 0;
        int leftPossible = Integer.MAX_VALUE;
        int rightMostMin = -1;
        int curMaxIdx = -1;
        int curMax = Integer.MIN_VALUE;
        int i = 0;
        
        for (int num : nums) {
            if (num < minK || num > maxK) {
                leftPossible = Integer.MAX_VALUE;
                rightMostMin = -1;
                curMax = Integer.MIN_VALUE;
            }

            if (num >= minK && num <= maxK) {
                leftPossible = Math.min(leftPossible, i);
                curMax = Math.max(curMax, num);
            }
            
            if (num == minK) {
                rightMostMin = i;
            }
            if (num == maxK) {
                curMaxIdx = i;
            }
            
            if (curMax == maxK && rightMostMin != -1 && leftPossible != Integer.MAX_VALUE) {
                
                res += (long) (Math.min(rightMostMin, curMaxIdx) - leftPossible + 1);
            }
            
            i++;
        }
        
        return res;
    }
}